File: eileenjang/LeetCode/73_Set_Matrix_Zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        int col = matrix.length, row = matrix[0].length;
        boolean firstRow = false;
        boolean firstCol = false;

        for (int i = 0; i < col; i++)
            if (matrix[i][0] == 0)
                firstCol = true;

        for (int j = 0; j < row; j++)
            if (matrix[0][j] == 0)
                firstRow = true;

        for (int i = 1; i < col; i++) {
            for (int j = 1; j < row; j++) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }

        for (int i = 1; i < col; i++) {
            for (int j = 1; j < row; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }

        if (firstCol) {
            for (int i = 0; i < col; i++)
                matrix[i][0] = 0;
        }

        if (firstRow) {
            for (int j = 0; j < row; j++)
                matrix[0][j] = 0;
        }
    }
}

// TC : O(M*N); SC : O(1)

File: yuridekim/LeetCode/73_Set_Matrix_Zeroes.py
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])

        row_list=[]
        col_list=[]

        for i in range(m):
            for j in range(n):
                if matrix[i][j]==0:
                    if i not in row_list:
                        row_list.append(i)
                    if j not in col_list:
                        col_list.append(j)

        """ runtime beats 62%     
        for i in range(m):
            for j in range(n):
                if i in row_list or j in col_list:
                    matrix[i][j]=0"""
                    
        # runtime beats 96%
        for idx in row_list:
            for j in range(n):
                matrix[idx][j]=0
        
        for idx in col_list:
            for i in range(m):
                matrix[i][idx]=0